### ISSUE TITLE
JsonSerializer polymorphic serialization and deserialization support

### ISSUE THREAD

**Issue by Symbai** [dotnet/runtime#30083]
EDIT see https://github.com/dotnet/runtime/issues/30083#issuecomment-861524767 for a finalized proposal.

<details>
<summary>
 Original proposal by @Symbai (click to view)
</summary>

```
public class FooBase
public class FooA : FooBase
public class FooB : FooBase
List<FooBase> SerializationObject = new List<FooBase> { new FooA(), new FooB() }
```
Serialized with Newtonsoft.JSON
```
JsonSerializerSettings Settings = new JsonSerializerSettings
{
    SerializationBinder = new KnownTypesBinder
    {
        KnownTypes = new List<Type> { typeof(FooA), typeof(FooB) }
    },
    TypeNameHandling = TypeNameHandling.Objects
};
List<FooBase> FooBases = new List<FooBase>() { new FooA(), new FooB() };
var json = JsonConvert.SerializeObject(FooBases, Settings);
```
will be:
`[{"$type":"FooA","NameA":"FooA","Name":"FooBase"},{"$type":"FooB","NameB":"FooB","Name":"FooBase"}]`

When using System.Text.Json.JsonSerializer to deserialize the json, FooA and FooB are both type of FooBase. Is there a way that JsonSerializer supports inheirited classes? How can I make sure the type will be the same and not the base class it inherits from?
</details>

**Comment by Symbai** [comment]
Why has this been removed from 5.0 roadmap?! We can currently not deserialize any POCO class, as well as any class which has an interface. This is definitely not a minor issue as both is widely used. Its a blocking issue for anyone currently using Newtonsoft. And its a known issue since about a year, so there was plenty of time. I'm sorry to say but I'm disappointed if this is something we need to wait another whole year when it ships with .NET 6 instead of .NET 5. And the bad taste also comes from the fact that serialization of these classes is supported already. So we have JSON support which only works into one direction only, it does not feel finished and ready to be shipped with .NET 5.

**Comment by lwardzala** [comment]
I've just discovered this thread and I decided to publish a solution I've been using in my projects.
It's an implementation of abstraction converter factory very easy to introduce.
It might be helpful till the issue will be closed :)
https://github.com/lwardzala/Json.Abstraction

**Comment by cocowalla** [comment]
I appreciate that the community are trying to backfill this missing functionality as a stopgap measure, but I have to join the chorus of voices here: it's crazy to release a JSON serialiser without polymorphic serialisation/deserialisation support, for an _object orientated language_, and then push people towards using it instead of Newtonsoft. I get it brings perf/allocation improvements, and that really is very much appreciated, but it also needs to actually be capable of real-world serialisation! :zany_face:

**Comment by bklooste** [comment]
Just because the language is polymorphic doesn't mean messages should be json and XML are NOT polymorphic . At best you  use a tagged union which the  language  interprets as polymorphic.  Many consumers of the message are not OO and its pretty bad practice IMHO to put this on the write and i have seen it cause tons of time wasting and issues.


KISS

**Comment by vpaulino** [comment]
So If I have an API based on contracts that share common properties I cannot make use of the language features is that it? I will need to replicate code? potencially increasing Technical Debt?

C# is ( until now ) an OOP language, so why are we building libraries that remove from us the usage of the basic language features? 

If the language nativelly has feature why are we creating that limitation?

Simple example:

```

public class ApiResult<T>  {

    public T Payload {get; set;}
    
}
```

```
public class PaginationApiResult<T> : ApiResult<IEnumerable<T>>
{
   public int Take {get; set; }
   public int Skip {get; set; }
  public long TotalRecords {get; set;}
}
```

Some other examples like Having a contract on my Frontend Api where I agregate data from multiple backends ( each with their own responsabilty) I want to have the capacity to minimize the number of requests to my Frontend API so reusing models helps me where I return a General Description of my Api Model and another contract where I return some detailed description of that model: 

```

public class PersonGeneralInfo {
 
public string CitizenId {get; set;}
public string FirstName {get; set;}
public string LastName {get; set;}
public DateTime BirthDate {get; set;}

}

public class EmployedPersonInformation : PersonGeneralInfo {
 
public DateTime StartedAt {get; set;}
public string CompanyName {get; set;}

}

public class PersonHealthInfo : PersonGeneralInfo {

public double Height {get; set; }
public double Weigth {get; set; }
public double BloodPresure {get; set; }

}

public class PersonLocationsInfo : PersonGeneralInfo {

public Address Home {get; set; }
public Address Work {get; set; }
public IEnumerable<Address> FavoritesRestaurants {get; set; }

}  
```

Many of these consumers are generated from swagger to typescript or C# applications or even Java Applications.

**Comment by Andrzej-W** [comment]
@bklooste you are only partially right. There are a lot of projects where people intentionally use C# on the server and on the client (for example Blazor) and these projects don't have to interact with components written in other languages. If it is necessary programmers can design non polymorphic interfaces but if it is not necessary we shouldn't have artificial limitations.

Every real business application uses decimal data type for monetary calculations. This data type is not supported in JavaScript and plenty of other languages. Does that mean we shouldn't use it in our programs? No, we need this data type and we don't care that some decimal numbers cannot be deserialized properly in JavaScript.

We use C# because it is an object oriented language and we also want to use polymorphism in JSON serialization/deserialization.

**Comment by Symbai** [comment]
Honestly, I'm getting tired of this discussion about the language and polymorphic. I totally got the point of "OOP is C# only" and "In the internet JSON isn't meant to support polymorphic". But here is the fact: If this JsonSerializer is meant to be a SERVER/INTERNET thing where OOP is out of scope.. then it should have been moved into System.Web, same as the existing [JavascriptSerializer](https://docs.microsoft.com/en-us/dotnet/api/system.web.script.serialization.javascriptserializer?view=netframework-4.8). Then nobody here would complain about missing polymorphic support. Because its god damn clear that this is a SERVER/INTERNET thing only.

But... it isn't in System.Web, so people complain about missing polymorphic support and they complain for a good reason. Even the old XmlSerializer supports polymorphic without too much additional code. 

So please stop and close this tab if you aren't interested in polymorphic support and simply accept the fact that many developers are and the way Microsoft has announced and design this JsonSerializer (Its not in System.Web and its many times compared against Newtonsoft.JSON which DOES support polymorphic) gives us the right to complain and ask for this missing feature.

**Comment by vpaulino** [comment]
Im not even talking about http ... durable functions for example use newtonsoft to store state between activities i think... probably due to some of the current limitations such as this one. I 100% agree that serialization should be considered to be used not only on the comunication world, but also to store state for example, ( REST is transfer state and databases is to store state)  and for that use case i would need to have a persistence model without OOP polymorphic feature? I think without Polymorphic support on Text.Json we will be reducing the usecases instead of being a more generic json serialization available to any use case.

**Comment by taspeotis** [comment]
I am publishing and consuming Kafka messages with their bodies in JSON format and I feel like the security concerns that blocked System.Text.Json from implementing polymorphic deserialization don't apply here.

I'm always happy to be educated about security but the way I see it is that the interactions between the Kafka cluster and my applications are closed to outside actors and the JSON is always going to be trusted.

So my preference would be for System.Text.Json to support this, and have it off by default to prevent people from shooting themselves in the foot if they do have untrusted input. But for trusted input scenarios I'm allowed to turn it on and use it.

**Comment by weitzhandler** [comment]
Here's [my attempt](https://stackoverflow.com/questions/58373915/serialize-objects-implementing-interface-with-system-text-json/64636093#64636093):

```c#
public class TypeMappingConverter<TType, TImplementation> : JsonConverter<TType>
  where TImplementation : TType
{
  [return: MaybeNull]
  public override TType Read(
    ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options) =>
      JsonSerializer.Deserialize<TImplementation>(ref reader, options);

  public override void Write(
    Utf8JsonWriter writer, TType value, JsonSerializerOptions options) =>
      JsonSerializer.Serialize(writer, (TImplementation)value!, options);
}
```

Usage:

```c#
var options =
   new JsonSerializerOptions 
   {
     Converters = 
     {
       new TypeMappingConverter<BaseType, ImplementationType>() 
     }
   };

JsonSerializer.Deserialize<Wrapper>(value, options);
```

---

Tests:

```c#
[Fact]
public void Should_serialize_references()
{
  // arrange
  var inputEntity = new Entity
  {
    References =
            {
                new Reference
                {
                    MyProperty = "abcd"
                },
                new Reference
                {
                    MyProperty = "abcd"
                }
            }
  };

  var options = new JsonSerializerOptions
  {
    WriteIndented = true,
    Converters =
    {
        new TypeMappingConverter<IReference, Reference>()
    }
  };

      var expectedOutput =
@"{
  ""References"": [
    {
      ""MyProperty"": ""abcd""
    },
    {
      ""MyProperty"": ""abcd""
    }
  ]
}";

  // act
  var actualOutput = JsonSerializer.Serialize(inputEntity, options);

  // assert
  Assert.Equal(expectedOutput, actualOutput);
}

[Fact]
public void Should_deserialize_references()
{
  // arrange

  var inputJson =
@"{
  ""References"": [
    {
      ""MyProperty"": ""abcd""
    },
    {
      ""MyProperty"": ""abcd""
    }
  ]
}";

  var expectedOutput = new Entity
  {
    References =
    {
      new Reference
      {
        MyProperty = "abcd"
      },
      new Reference
      {
        MyProperty = "abcd"
      }
    }
  };

  var options = new JsonSerializerOptions
  {
    WriteIndented = true
  };

  options.Converters.AddTypeMapping<IReference, Reference>();

  // act
  var actualOutput = JsonSerializer.Deserialize<Entity>(inputJson, options);

  // assert
  actualOutput
      .Should()
      .BeEquivalentTo(expectedOutput);
}


public class Entity
{
  ICollection<IReference>? _References;
  public ICollection<IReference> References
  {
    get => _References ??= new HashSet<IReference>();
    set => _References = value;
  }
}

public interface IReference
{
  public string? MyProperty { get; set; }
}

public class Reference : IReference
{
  public string? MyProperty { get; set; }
}
```

**Comment by vpaulino** [comment]
I have implemented something about that available [here] on one of my github repos (https://github.com/vpaulino/BlogSamples/tree/main/PolymorphicTextJsonConverter). Im writing something about that on my blog ( I do not write that much but the important stuff I think is in that repo )

**Comment by SCLDGit** [comment]
Something that has yet to have been brought up is people with no web expectations at all who are simply using the JSON format as the excellent human-readable, compact interchange format that it is for use cases totally outside of web traffic. Limiting our ability to handle polymorphic data, like that from classes inherited from interfaces, is silly and shortsighted.

**Comment by robkeimig** [comment]
This is the only thing stopping us from moving to System.Text.Json. We _really_ want to, but we have to have polymorphic support both ways. It might not be "as intended" or "best practice", but we use JSON for serializing most of our business state to/from durable storage. When you are working with incredibly dense object graphs that can change on a daily basis (over 1000 properties across 100+ types), you don't have the luxury of mapping one-property-per-column to your persistence layers and rolling migrators each time. At least not at our scale. JSON serialization of our business models has proven to be, by orders of magnitude, the most sustainable way to manage a huge aspect of our codebase. As a result, we prefer to have full control over serialization at all levels. We have inheritance hierarchies 4-5 types deep that are serializing like a dream w/ Newtonsoft's `TypeNameHandling.Auto`. No special attributes, helper properties, interfaces, or converter nonsense required.  

Due to the level of integration we have with Newtonsoft and the terabytes of existing JSON state we have on customer servers, we will probably have to roll something in-house that 1:1 compatible. We will probably never be able to use this library. Despite this, I like to bring my case to bear because I am sure there are others who use JSON as a storage format for complex data sets, and of these there are probably some who would love to be able to use polymorphism for better modeling in code.

Certainly for web-based usage, a polymorphic deserialization proposal is suspect at best. I would say even for an internal application, allowing the client to drive type selection is not a good idea.  We do not want to use any form of polymorphic serialization on our AspNetCore hosts. Only for private or other trusted back-end code that reads/writes business entities to/from various byte streams.

**Comment by IngemarNilsson** [comment]
> I've just discovered this thread and I decided to publish a solution I've been using in my projects.
> It's an implementation of abstraction converter factory very easy to introduce.
> It might be helpful till the issue will be closed :)
> https://github.com/lwardzala/Json.Abstraction

Great, thanks!

**Comment by IngemarNilsson** [comment]
> I've just discovered this thread and I decided to publish a solution I've been using in my projects.
> It's an implementation of abstraction converter factory very easy to introduce.
> It might be helpful till the issue will be closed :)
> https://github.com/lwardzala/Json.Abstraction

Thanks, this one did it for me!

**Comment by GrabYourPitchforks** [comment]
\[security hat on\]

Friendly reminder - unrestricted polymorphic deserialization is a remote code execution vector. Using Newtonsoft.Json with TypeNameHandling set to any value other than _TypeNameHandling.None_ is insecure unless you have also configured a binder. We hold ourselves to that same bar: Microsoft first-party code which uses Newtonsoft.Json as a serializer is forbidden from touching the TypeNameHandling property. The exception process is intentionally onerous, even with a binder in place.

The safe way to support this is to have a fixed mapping of allowed types. In order to be secure, this mapping must be provided _before_ deserialization takes place. For example:

```json
{
    "$type": "Dog",
    "Name": "Fido",
    "Age": 8,
    "Breed": "Terrier"
}
```

You can imagine the deserializer having access to a `Dictionary<string, Type>`, where the value of the `$type` property is used as a lookup key into this dictionary, and the `Type` value is used as the actual type to be instantiated and to have its members populated.

```cs
// assume the allowed mapping is specified via the 'options' parameter below

Options options = new Options();
options.AllowedMappings["Dog"] = typeof(Contoso.Models.Dog);

Animal animal = Deserialize<Animal>(payload, options);
Console.WriteLine(animal.GetType()); // prints 'Contoso.Models.Dog'
```

Importantly:

* `Type.GetType(string)` and similar APIs must __never under any circumstance__ be called with untrusted input. This implies that feeding the value of the `$type` field into `Type.GetType` is disallowed.

* The caller must pre-populate the mapping dictionary _before_ the call to deserialize, or there must be some other deterministic and documented way for the deserializer to discover the set of allowed types based solely on statically available type information from the caller. (This can also be provided by a source analyzer or other AOT mechanism.)

* The mapping dictionary should avoid being clever in attempting to support things like open generics or collections (e.g., `List<>`, `T[]`). This may allow unintended recursive nesting, which can re-introduce security holes into the application.

These guidelines will drive any implementation of a polymorphism-enabled deserializer provided by the library. This is required by the security review process Microsoft follows during development.

\[security hat off - please have a pleasant day\]

**Comment by lwardzala** [comment]
> [security hat on]
> 
> Friendly reminder - unrestricted polymorphic deserialization is a remote code execution vector. Using Newtonsoft.Json with TypeNameHandling set to any value other than _TypeNameHandling.None_ is insecure unless you have also configured a binder. We hold ourselves to that same bar: Microsoft first-party code which uses Newtonsoft.Json as a serializer is forbidden from touching the TypeNameHandling property. The exception process is intentionally onerous, even with a binder in place.
> 
> The safe way to support this is to have a fixed mapping of allowed types. In order to be secure, this mapping must be provided _before_ deserialization takes place. For example:
> 
> ```json
> {
>     "$type": "Dog",
>     "Name": "Fido",
>     "Age": 8,
>     "Breed": "Terrier"
> }
> ```
> 
> You can imagine the deserializer having access to a `Dictionary<string, Type>`, where the value of the `$type` property is used as a lookup key into this dictionary, and the `Type` value is used as the actual type to be instantiated and to have its members populated.
> 
> ```cs
> // assume the allowed mapping is specified via the 'options' parameter below
> 
> Options options = new Options();
> options.AllowedMappings["Dog"] = typeof(Contoso.Models.Dog);
> 
> Animal animal = Deserialize<Animal>(payload, options);
> Console.WriteLine(animal.GetType()); // prints 'Contoso.Models.Dog'
> ```
> 
> Importantly:
> 
> * `Type.GetType(string)` and similar APIs must **never under any circumstance** be called with untrusted input. This implies that feeding the value of the `$type` field into `Type.GetType` is disallowed.
> * The caller must pre-populate the mapping dictionary _before_ the call to deserialize, or there must be some other deterministic and documented way for the deserializer to discover the set of allowed types based solely on statically available type information from the caller. (This can also be provided by a source analyzer or other AOT mechanism.)
> * The mapping dictionary should avoid being clever in attempting to support things like open generics or collections (e.g., `List<>`, `T[]`). This may allow unintended recursive nesting, which can re-introduce security holes into the application.
> 
> These guidelines will drive any implementation of a polymorphism-enabled deserializer provided by the library. This is required by the security review process Microsoft follows during development.
> 
> [security hat off - please have a pleasant day]

Thanks for the guideline! I'll try to review and adapt that in my solution to make it easier to implement and safer as well.

**Comment by BreyerW** [comment]
@GrabYourPitchforks would be ok to provide polymorphism support via something like `TypeResolver` that mimics `ReferenceResolver` but for `$type` metadata? This way different ploymorphism mechanism could be provided depending on security requirements (eg not every app holds sensitive data or some live in isolated enviroments meaning data is always trusted)

**Comment by jeremyVignelles** [comment]
I think that @Timovzl is really spot on here : https://github.com/dotnet/runtime/issues/30083#issuecomment-534078476

Unless I'm missing something, there is a big prerequisite that is currently missing: `Utf8JsonReader` is a forward-only reader, and the `$type` or whatever discrimination property must be the first to be read, otherwise you'll be missing some properties that you can't catch later, unless you buffered them. There is no guarantee that it will be the case for every use case, especially when consuming APIs, like OpenAPI web services.

It's commonn in other languages like typescript to have type unions (like `A | B`) instead of proper inheritance, and the OpenAPI specification itself offers various mechanisms to deal with [composition and inheritance](https://swagger.io/specification/#composition-and-inheritance-polymorphism). There's no strong rule however that the discriminator is always first, and we should assume that it might not be the case, and fixing the input json is likely not an option.

What I mean is : Yes, `$type` deserialization should be possible for compatibility with Newtonsoft APIs or other things, but it is a use case inside a bigger feature, and I'm not even sure it should be supported out of the box. If the tools are there, one could, with the help of samples, write their own converters for the case/create a library that could be reused by everyone.

I've been diving in the `A | B` issue which is cructial for proper OpenAPI spec support. I created a `TypeUnion<A, B>` type with a `JsonConverter` that would try to deserialize as either A or B. Obviously, I faced the same issue as some here, and had to deserialize/reserialize as others have pointed here.

You can follow my work in progress in [the Ooak repository](https://github.com/jeremyVignelles/ooak), and especially [this piece of code](https://github.com/jeremyVignelles/ooak/blob/9c7293a9acd98f61702d6fdb05d02c92ef358ad2/Ooak/SystemTextJson/OoakSystemTextJsonConverter.cs#L65-L70) in the converter.

So now that the issue is described what options would we have to solve it?

1. Make the Utf8Reader "rewindable" or save/restore its internal state somehow, like a snapshot.
2. Create a new Utf8Reader each time we want to "try" a branch.
3. Any other idea?

# Make the Utf8Reader "rewindable" or save/restore its internal state somehow, like a snapshot.

This doesn't sound like a good idea to mess with internal state, but that might be a straightforward approach. I'm sure there are good reasons not to do it that way...

# Create a new Utf8Reader each time we want to "try" a branch.

There is already a JsonReaderState available, that represents the internal state of the reader, but that requires the input `ReadOnlySpan<byte>` which we don't have access to, if I'm not missing anything. Maybe if we had access to that input, we could create another reader with the current state and data from the current position each time we need to "branch" the code.

I don't see any hard blockers since the source is ReadOnly, and it doesn't affect the current reader, but it's harder to implement a serializer since we must advance the original reader at the end.

# Closing

I'm pretty sure I don't have the full picture, and specifically, I'm not sure about how this plays with "chunked" json (if the json data comes in multiple `ReadOnlySpan<byte>`, which is possible with pipes...

**Comment by GrabYourPitchforks** [comment]
> Make the Utf8Reader "rewindable" or save/restore its internal state somehow, like a snapshot.

This is already possible today. It's a struct with all relevant state data present as instance fields, so you can copy the struct by value. Example:

```cs
Utf8JsonReader reader = GetReader();

/* do some work with reader */

Utf8JsonReader snapshot = reader; // assign a copy to a local

/* do more with with 'reader' */

reader = snapshot; // rewinds the reader back to what it was when the snapshot was taken
```

**Comment by jeremyVignelles** [comment]
Thanks for your answer @GrabYourPitchforks . I tried the solution of breaking the `internal` to get the `OriginalSpan` to see if there's something we could do in the converter.
Unfortunately, it seems that the JsonConverter receives a reader that has already consumed the first token to have its Value fields initialized.
This means I can't use JsonSerializer.Deserialize on my new reader initialized with the state as it's already past the start of the object.
Am I missing something?

**Comment by jeremyVignelles** [comment]
So based on @GrabYourPitchforks , I modified my converter and it seems pretty neat to just copy the reader. See this code for example: https://github.com/jeremyVignelles/ooak/blob/3cb644e002f8f58fd91b1144d99c9b58194594f2/Ooak/SystemTextJson/OoakSystemTextJsonConverter.cs#L86

Based on that attempt, I quickly hacked a custom converter to attempt a polymorphic deserialization, and it turns out it works pretty easily. I hosted the code [here](https://github.com/jeremyVignelles/SystemTextJsonPolymorphicDeserialization/)

Here is the converter:

```cs
using System;
using System.Collections.Generic;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace SystemTextJsonPolymorphicDeserialization
{
    public class PolymorphicConverter : JsonConverter<BaseModel>
    {
        public readonly Dictionary<string, Type> TypeMapping = new Dictionary<string, Type> {
            { "Bar", typeof(BarModel) },
            { "Foo", typeof(FooModel) }
        };

        public override BaseModel? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
        {
            if(reader.TokenType != JsonTokenType.StartObject)
            {
                throw new JsonException("Expected start of object");
            }
            var typeReader = reader;
            Type? deserializedType = null;
            while(typeReader.Read())
            {
                if(typeReader.TokenType == JsonTokenType.PropertyName && typeReader.GetString() == "Type")
                {
                    if(!typeReader.Read())
                    {
                        throw new JsonException($"Unable to read.");
                    }
                    var typeName = typeReader.GetString();
                    if (typeName is null)
                    {
                        throw new JsonException($"Unable to read the type name.");
                    }
                    if(!this.TypeMapping.TryGetValue(typeName, out deserializedType))
                    {
                        throw new JsonException($"Invalid type in json : {typeName}");
                    }
                    break;
                }
                else if(typeReader.TokenType == JsonTokenType.StartArray || typeReader.TokenType == JsonTokenType.StartObject)
                {
                    typeReader.Skip();
                }
            }

            if(deserializedType == null)
            {
                throw new JsonException($"Key 'Type' not found.");
            }

            return (BaseModel?)JsonSerializer.Deserialize(ref reader, deserializedType, options);
        }

        public override void Write(Utf8JsonWriter writer, BaseModel value, JsonSerializerOptions options)
        {
            throw new NotImplementedException();
        }
    }
}
```

I really like the way I could branch the reader to detect a given propertyName `Type` in my example, but might as well be `$type` or whatever. It's way better than parsing twice in a JsonDocument or other hacks I've seen.

But I'm under the impression that I'm missing something really simple. If it's that simple, how come I'm the first one to discover this? Isn't there any caveat I'm ignoring here, like chunked json (ReadOnlySegments...) or something ?

If it really works, it could be pretty easy to code an API around that to register known subclasses of a given type, but defining the API shape might be tricky

**Comment by ahsonkhan** [comment]
> It's way better than parsing twice in a JsonDocument or other hacks I've seen.

:)

> But I'm under the impression that I'm missing something really simple. If it's that simple, how come I'm the first one to discover this?

Others have posed similar questions. It is understandable that some folks miss that `Utf8JsonReader` is a struct with value semantics and not a `class`, but as you observed, it makes resetting or taking a snapshot quite easy. The serializer and JsonDocument internally use this approach some times too:
https://stackoverflow.com/a/59744873/12509023
![image](https://user-images.githubusercontent.com/6527137/102685232-0896d500-4194-11eb-94d1-e37507732a42.png)
https://github.com/dotnet/runtime/blob/6dd7b2df5098a61514b13114cb5bd682dff5c71c/src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs#L353-L354

> Isn't there any caveat I'm ignoring here, like chunked json (ReadOnlySegments...) or something ?

No, I don't think so. You don't seem to be doing anything with the reader that would break if  the input was chunked as a `ReadOnlySequence`. In most cases, as long as you are using standard public methods on the `Utf8JsonReader`, whether the input was a contiguous `ROSpan` or split in chunks as a `ROSequence` shouldn't be a concern you need to worry about. The reader deals with that for you since almost all the APIs on it are agnostic to that detail. You could try adding some test cases to verify for yourself. Outside of the context of the serializer (i.e. when directly using the reader), you would have to be careful about one other fact, which is processing partial or incomplete JSON data, since the reader is re-entrant and supports incomplete input. However, in the context of JsonSerializer and within the `JsonConverter` you write, that is not a concern because the serializer reads ahead to ensure the JSON you are deserializing contains at least enough data to complete the current object successfully, before calling your `JsonConverter`. So, the `reader.Read()` method within a `JsonConverter` will not return false.
https://github.com/dotnet/runtime/blob/6dd7b2df5098a61514b13114cb5bd682dff5c71c/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverter.ReadAhead.cs#L43-L48

https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-converters-how-to?pivots=dotnet-5-0#steps-to-follow-the-basic-pattern from https://github.com/dotnet/docs/issues/19654
> Override the Read method to deserialize the incoming JSON and convert it to type T. Use the Utf8JsonReader that is passed to the method to read the JSON. You don't have to worry about handling partial data, as the serializer passes all the data for the current JSON scope. So it isn't necessary to call Skip or TrySkip or to validate that Read returns true.

**Comment by jeremyVignelles** [comment]
> However, in the context of JsonSerializer and within the JsonConverter you write, that is not a concern because the serializer reads ahead to ensure the JSON you are deserializing contains at least enough data to complete the current object successfully, before calling your JsonConverter. So, the reader.Read() method within a JsonConverter will not return false.

This is what I needed, thanks for the explanation. That makes sense, because I never saw any converter example dealing with incomplete data.

I didn't know there was an example for polymorphic deserialization, but as pointed in your stackoverflow thread, the discriminator is outside the data being read, which is not the case for the objects serialized with Newtonsoft for example, so the converter example I posted above could be really useful.

**Comment by bklooste** [comment]
> @bklooste you are only partially right. There are a lot of projects where people intentionally use C# on the server and on the client (for example Blazor) and these projects don't have to interact with components written in other languages. If it is necessary programmers can design non polymorphic interfaces but if it is not necessary we shouldn't have artificial limitations.

I have lost count of how often i have heard this argument .. im ancient now and every time management or some new dev team changes the back end to Java / go or front end to  js this has ended up costing a mint . Its hard language based coupling. ( Why are you using an interface again... OO is dying it doesnt need to die but dont distribute it)  

Not to mention the expensive C#  cost when the deserializer breaks because some rarely used thing breaks and these are EXPENSIVE, painful to find and fix bugs often needing contracts changed across many services..  What do you gain you save a handful of  lines in a SIMPLE converter at a cost of hidden complexity and expensive bugs ?  Note it is NOT datetime / decimal conversion that is the issue all serializes do that which is fine ( as it does not change on wire representation) its only hacking polymorphism into  json when its not  designed for it ( and XML was the same) .   Typeof has also been used its better but again its not worth it leading to bugs because it gets omitted. 

You still need polymorphism sometimes on the wire but do it like every other modern language eg a dev managed tagged union. 

I admit there are a few corner cases within a small app but on the wire/ in messages for larger things its bad practice with high maintenance. I used to do it 15 years ago but since then its not worth it. Especially with multi language micro services growing. 

Anyway each to his own  but i though i would share my experience.

**Comment by eiriktsarpalis** [comment]
## Propoposed Design (polymorphism using type discriminators)

This feature allows users to opt in to polymorphic serialization for a given type by associating string identifiers with particular subtypes in the hierarchy. These identifiers are written to the wire so this brand of polymorphism is roundtrippable.

At the core of the design is the introduction of `JsonKnownType` attribute that can be applied to type hierarchies like so:
```csharp
[JsonKnownType(typeof(Derived1), "derived1")]
[JsonKnownType(typeof(Derived2), "derived2")]
public class Base
{
    public int X { get; set; }
}

public class Derived1 : Base
{
    public int Y { get; set; }
}

public class Derived2 : Base
{
    public int Z { get; set; }
}
```
This allows roundtrippable polymorphic serialization using the following schema:
```csharp
var json1 = JsonSerializer.Serialize<Base>(new Derived1()); // { "$type" : "derived1", "X" : 0, "Y" : 0 }
var json2 = JsonSerializer.Serialize<Base>(new Derived2()); // { "$type" : "derived2", "X" : 0, "Z" : 0 }

JsonSerializer.Deserialize<Base>(json1); // uses Derived1 as runtime type
JsonSerializer.Deserialize<Base>(json2); // uses Derived2 as runtime type
```
Alternatively, users can specify known type configuration using the
`JsonSerializerOptions.TypeDiscriminatorConfigurations` property:
```csharp
public class JsonSerializerOptions
{
    public IList<TypeDiscriminatorConfiguration> TypeDiscriminatorConfigurations { get; }
}
```
which can be used as follows:
```csharp
var options = new JsonSerializerOptions
{
    TypeDiscriminatorConfigurations =
    {
        new TypeDiscriminatorConfiguration<Base>()
          .WithKnownType<Derived1>("derived1")
          .WithKnownType<Derived2>("derived2")
    }
};
```
or alternatively
```csharp
var options = new JsonSerializerOptions
{
    TypeDiscriminatorConfigurations =
    {
        new TypeDiscriminatorConfiguration(typeof(Base))
          .WithKnownType(typeof(Derived1), "derived1")
          .WithKnownType(typeof(Derived2), "derived2")
    }
};
```

## Proposed API

```diff
namespace System.Text.Json
{
+    [AttributeUsageAttribute(AttributeTargets.Class | AttributeTargets.Interface, AllowMultiple = true, Inherited = false)]
+    public partial class JsonKnownTypeAttribute : JsonAttribute
+    {
+        public JsonKnownTypeAttribute(Type subtype, string identifier) { }
+        public string Identifier { get { throw null; } }
+        public System.Type Subtype { get { throw null; } }
+    }

+    public class TypeDiscriminatorConfiguration : IReadOnlyCollection<KeyValuePair<System, string>>
+    {
+        public TypeDiscriminatorConfiguration(Type baseType);
+        public System BaseType { get; }
+        public TypeDiscriminatorConfiguration WithKnownType(Type derivedType, string identifier) { }
+    }

+    public class TypeDiscriminatorConfiguration<TBaseType> : TypeDiscriminatorConfiguration 
+        where TBaseType : class
+    {
+        public TypeDiscriminatorConfiguration();
+        public TypeDiscriminatorConfiguration<TBaseType> WithKnownType<TDerivedType>(string identifier) 
+            where TDerivedType : TBaseType { }
+    }

    public class JsonSerializerOptions
    {
+        public IList<TypeDiscriminatorConfiguration> TypeDiscriminatorConfigurations { get; }
    }
}
```

## Open Questions

### Customizing the type discriminator property name

The property name is currently hardcoded to `$type` but it stands to reason that users might need to customize this, potentially using a special attribute:
```csharp
[JsonTypeDiscriminatorPropertyName("_case")]
[JsonKnownType(typeof(MyDerivedType), "derived")]
public class MyPoco { }
public class MyDerivedtype : MyPoco { }

JsonSerializer.Serialize(new MyDerivedType()); // { "_case" : "derived" }
```

### Using type discriminators that are not strings

We might want to allow specifying non-string type identifiers:
```csharp
[JsonKnownType(typeof(MyDerivedtype), 0)]
[JsonKnownType(typeof(MyOtherDerivedtype), 1)]
public class MyPoco { }
public class MyDerivedtype : MyPoco { }
public class MyOtherDerivedtype : MyPoco { }

JsonSerializer.Serialize(new MyDerivedType()); // { "_case" : 0 }
```
The only valid identifiers would be strings and integers and it should not be possible for a given type hierarchy to mix and match identifier types.

### Semantics

The type discriminator semantics could be implemented following two possible alternatives, which for the purposes of this document I will be calling "strict mode" and "lax mode". Each approach comes with its own sets of trade-offs.

#### Strict mode

"Strict mode" requires that any runtime type used during serialization must explicitly specify a type discriminator. For example:
```csharp
[JsonKnownType(typeof(Derived1),"derived1")]
[JsonKnownType(typeof(Derived2),"derived2")]
public class Base { }

public class Derived1 : Base { }
public class Derived2 : Base { }
public class Derived3 : Base { }

public class OtherDerived1 : Derived1 { }

JsonSerializer.Serialize<Base>(new Derived1()); // { "$type" : "derived1" }
JsonSerializer.Serialize<Base>(new Derived2()); // { "$type" : "derived2" }
JsonSerializer.Serialize<Base>(new Derived3()); // throws NotSupportedException
JsonSerializer.Serialize<Base>(new OtherDerived1()); // throws NotSupportedException
JsonSerializer.Serialize<Base>(new Base()); // throws NotSupportedException
```
Any runtime type that is not associated with a type discriminator will be rejected, including instances of the base type itself. This approach has a few drawbacks:

* Does not work well with open hierarchies: any new derived types will have to be explicitly opted in.
* Each runtime type must use a separate type identifier.
* It is not possible to specify identifiers for subtypes that are interfaces or abstract classes.

#### Lax mode

"Lax mode" as the name suggests is more permissive, and runtime types without discriminators are serialized using the nearest type ancestor that does specify a discriminator. Using the previous example:
```csharp
[JsonKnownType(typeof(Derived1),"derived1")]
[JsonKnownType(typeof(Derived2),"derived2")]
public class Base { }

public class Derived1 : Base { }
public class Derived2 : Base { }
public class Derived3 : Base { }

public class OtherDerived1 : Derived1 { }

JsonSerializer.Serialize<Base>(new Derived1()); // { "$type" : "derived1" }
JsonSerializer.Serialize<Base>(new Derived2()); // { "$type" : "derived2" }
JsonSerializer.Serialize<Base>(new Derived3()); // { } serialized as `Base`
JsonSerializer.Serialize<Base>(new OtherDerived1()); // { "$type" : "derived1" } inherits schema from `Derived1`
JsonSerializer.Serialize<Base>(new Base()); // { } serialized as `Base`
```
This approach is more flexible and supports interface and abstract type hierarchies:
```csharp
[JsonKnownType(typeof(Foo), "foo")]
[JsonKnownType(typeof(IBar), "bar")]
public interface IFoo { }
public abstract class Foo : IFoo { }
public interface IBar : IFoo { }

public class FooImpl : Foo {}

JsonSerializer.Serialize<IFoo>(new FooImpl()); // { "$type" : "foo" }
```
However it does come with its own set of problems:
```csharp
[JsonKnownType(typeof(Foo), "foo")]
[JsonKnownType(typeof(IBar), "bar")]
public interface IFoo { }
public class Foo : IFoo { }
public interface IBar : IFoo { }

public Baz : Foo, IBar { }

JsonSerializer.Serialize<IFoo>(new Baz()); // diamond ambiguity, could either be "foo" or "bar",
                                           // throws NotSupportedException.
```

[Polymorphism Draft Design Document](https://github.com/dotnet/designs/pull/226)
[Prototype Implementation](https://github.com/dotnet/runtime/pull/53882)

**Comment by eiriktsarpalis** [comment]
We have decided to move this feature to .NET 7, since implementing it requires substantial [infrastructural refactorings](https://github.com/dotnet/runtime/pull/53882#issue-665078110) in the core serialization logic. Since we are fairly late in the .NET 6 development cycle, introducing the changes at this point carries risk of regressions.

**Comment by BreyerW** [comment]
@eiriktsarpalis 

I would like to ask if mirroring ReferenceResolver API for polymorphic support with appropiate tweaks was considered? This would allow providing both newtonsoft compatible and secure with type discriminators implementations of polymorphic support and also other flavours. Secure implementation could be provided by default in bcl and newtonsoft provided in guide to migrate but not in bcl to discourage wide use of it

**Comment by eiriktsarpalis** [comment]
@BreyerW you mean abstracting the polymorphism resolution logic behind an abstract class? What specific scenario do you have in mind that would be unlocked by such a design versus what is being proposed? Would you be able to sketch an API design for the resolver type?

Note that we are intentionally not making this design compatible with (or as capable as) Json.NET polymorphism, out of security considerations related to reading type names from the wire.

**Comment by lwardzala** [comment]
@eiriktsarpalis
So what in case of layered architecture when we have a strict hierarchy like in Sitecore framework or just in some applications?
Couldn't the relations be registered as it is in MongoDB driver or as I implemented here (https://github.com/lwardzala/Json.Abstraction)?
I'm aware that the performance would be much better with your proposed design but still I would like to avoid two-direction dependency flow.

**Comment by eiriktsarpalis** [comment]
> I'm aware that the performance would be much better with your proposed design but still I would like to avoid two-direction dependency flow.

I wouldn't say it's more of a security consideration rather than a performance one: the feature will not automatically support open hierarchies (since that requires serializing and deserializing the type name from the wire). Individual derived types will have to be opted in manually, akin to the [`KnownTypes` attribute](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.serialization.knowntypeattribute?view=net-5.0) in DataContractSerializer.

If circular dependency between base and derived type is not desirable (or indeed possible, in cases where the hierarchy spans multiple assemblies) you can register an equivalent `TypeDiscriminatorConfiguration` instance in your `JsonSerializerOptions`, as described in the proposal.

**Comment by lwardzala** [comment]
Ah yes,
`JsonSerializerOptions.TypeDiscriminatorConfigurations` looks great.
I've just misread that part :)

**Comment by BreyerW** [comment]
@eiriktsarpalis 

i was thinking something like this:

```csharp
abstrac class TypeResolver{

//called when $type is read
public abstract Type ResolveType(string);

//called during write, bool indicates if it should be written at all to support type discriminators. 
//Alternatively null on return could be used to indicate that type metadata should NOT be emitted. Then bool isnt needed. 
public abstract string GetTypeMetadata(Type, out bool);

}
```

Since `TypeResolver` isnt supposed to be factory for real objects (i think) smaller abstract class than `ReferenceResolver` like the one presented should suffice.

As for scenarios unblocked it is more about future proofing. I have only skimmed current draft but current design seems to be a bit rigid here and there, mostly around type discriminators (i may be wrong on that though) which causes me a bit of anxiety.

Eg. what if some obscure security hole is discovered in type discriminator? Not necessarily in general/pattern idea - although that is also possible even if unlikely - but in your specific algorithm? Then we have to wait until you fix hole. With abstract approach it would be fixable on our end if necessary. Also possibly someone may wanted different trade offs in perf (eg cpu vs memory)?  Also possible bugs in threaded scenarios...

They are all very abstract and probably not very likely but still it is something im slighty worried about

**Comment by FiniteReality** [comment]
> ```cs
> abstract class TypeResolver{
> 
> //called when $type is read
> public abstract Type ResolveType(string);
> 
> //called during write, bool indicates if it should be written at all to support type discriminators. 
> //Alternatively null on return could be used to indicate that type metadata should NOT be emitted. Then bool isnt needed. 
> public abstract string GetTypeMetadata(Type, out bool);
> 
> }
> ```

I think this API is only good surface deep. If anything, I think it'll encourage users to just write:
```cs
class MyVeryInsecureTypeResolver : TypeResolver
{
    public override Type ResolveType(string type) => Type.GetType(type);
    public override string GetTypeMetadata(Type type) => type.AssemblyQualifiedName;
}
```
Which, as we know from BinaryFormatter, is very insecure :sweat_smile:

If this is the approach that gets chosen, it needs to come with an analyzer that screams as hard as it can if people do something like that, or it needs to come with a lot of default implementations which people are encouraged to use.

**Comment by FiniteReality** [comment]
> The only valid identifiers would be strings and integers and it should not be possible for a given type hierarchy to mix and match identifier types.

I still think supporting types inheriting from `Enum` would be advantageous here. As a workaround, I presume the advice would be to cast to int? Something along the lines of this:
```cs
public enum MyJsonTypes
{
    TypeOne,
    TypeTwo
    // etc.
}

[JsonKnownType(typeof(TypeOne), (int)MyJsonTypes.TypeOne)]
[JsonKnownType(typeof(TypeTwo), (int)MyJsonTypes.TypeTwo)]
public class BaseClass { }

public class TypeOne : BaseClass { }
public class TypeTwo : BaseClass { }
```

**Comment by eiriktsarpalis** [comment]
> I still think supporting types inheriting from Enum would be advantageous here. As a workaround, I presume the advice would be to cast to int?

We could support accepting enums directly, but that would require changing the attribute constructor signature to accept `object` identifiers:
```csharp
public partial class JsonKnownTypeAttribute : JsonAttribute
{
     public JsonKnownTypeAttribute(Type subtype, object identifier) { }
     public object Identifier { get { throw null; } }
     public System.Type Subtype { get { throw null; } }
}
```
but that would enable passing invalid identifier types at compile time such as flots or `Type` instances. I was thinking we could instead enforce a degree of type safety by exposing two constructors:
```csharp
public partial class JsonKnownTypeAttribute : JsonAttribute
{
     public JsonKnownTypeAttribute(Type subtype, string identifier) { }
     public JsonKnownTypeAttribute(Type subtype, int identifier) { }
     public object Identifier { get { throw null; } }
     public System.Type Subtype { get { throw null; } }
}
```

**Comment by BreyerW** [comment]
@eiriktsarpalis  do you think generics in attributes would enable enums in safe manner? Afaik generics in attributes are already implemented just not merged for c# due to bug in net framework but c# team recently admitted they no longer consider it a showstopper because enough time lapsed since net framework stopped being developed further. They probably just need a nudge and this scenario might be that nugde (also it is scheduled for net 7 now meaning timeframe for generic in attributes is up to c# 11 not just c# 10 which would be rather short timeframe from now)

**Comment by eiriktsarpalis** [comment]
@BreyerW a generic attribute would not prevent invalid specializations, e.g. `new JsonKnownTypeAttribute<Type>(typeof(int), typeof(int))` so it would be largely equivalent to accepting an `object` identifier.

**Comment by TylerBrinkley** [comment]
I think @BreyerW was thinking of something like this but I'm not sure it's a great experience even if generic attributes were already available.

```c#
public partial class JsonKnownTypeEnumAttribute<TEnum> : JsonKnownTypeAttribute
    where TEnum : struct, Enum
{
     public JsonKnownTypeEnumAttribute(Type subtype, TEnum identifier) { }
}

public partial class JsonKnownTypeAttribute : JsonAttribute
{
     public JsonKnownTypeAttribute(Type subtype, string identifier) { }
     public JsonKnownTypeAttribute(Type subtype, int identifier) { }
     internal JsonKnownTypeAttribute(Type subType, object identifier) { }
     public object Identifier { get { throw null; } }
     public System.Type Subtype { get { throw null; } }
}
```

**Comment by eiriktsarpalis** [comment]
Got it. It seems like something users could do by inheriting from the base attribute type if they really needed it.

**Comment by BreyerW** [comment]
@TylerBrinkley actually you dont even need Enum in name it is possible to have types with same name but differ by existence of generic (checked it in sharplab) making experience a bit better although i guess @eiriktsarpalis  is right and we could do that ourselves

**Comment by bklooste** [comment]
Yeah generics modfied by types in different assemblies is always fun when you need to create them ( Assembly.load and gettype under the covers of Activator.GetInstance etc) and tends to be slow for lots of types /objects.   I dont think you can easily have the new () constraint on any type unless you do it yourself but should be looked at if you go that path.

**Comment by LarsKemmann** [comment]
@eiriktsarpalis I'm not familiar enough with the details of the serialization/type converter stack... will the proposed design support nested polymorphic types? E.g.:
```c#
[JsonKnownType(typeof(Derived1),"derived1")]
[JsonKnownType(typeof(Derived2),"derived2")]
public record Base();
public record Derived1(int X, NestedBase NestedValue) : Base();
public record Derived2(string Y) : Base();

[JsonKnownType(typeof(NestedDerived1),"nestedderived1")]
[JsonKnownType(typeof(NestedDerived2),"nestedderived2")]
public record NestedBase(string Foo);
public record NestedDerived1(string Foo, double Z) : NestedBase(Foo);
public record NestedDerived2(string Foo) : NestedBase(Foo);
```

Could I round-trip a `Derived1` with a `NestedValue` of type `NestedDerived1`, and will the value of `Z` round-trip as well?

**Comment by eiriktsarpalis** [comment]
Hi @LarsKemmann, yes that particular scenario would certainly be supported.

**Comment by tthiery** [comment]
Please make sure to support configuring the type discriminator property name. "$type" is very common but not universal.

Thanks for picking this up. I wrote my own one with a bad performance pretty much like this one but with the caveat of the lacking infrastructure.

**Comment by IFYates** [comment]
While this change looks great for a round-trip solution, does it mean that there's no chance for a simple option on the serializer so that serialized output resolves polymorphic values when there's no need for deserializing later?

**Comment by tthiery** [comment]
I want to nit pick on "roundtrip": polymorphic deserialization is about receiving input which has a type attribute. Roundtrip or .net to .net is a tiny percentage of the demand.

**Comment by Ilchert** [comment]
I [wrote](https://gist.github.com/Ilchert/4854a20f790ca963d1ab17d99433c7da) a simple converter to add polymorphic serialization. It is a stub to add polymorphic serialization without build in support.

**Comment by Ilchert** [comment]
@IFYates you can [write](https://gist.github.com/Ilchert/4854a20f790ca963d1ab17d99433c7da) your own converter factory to handle polymorphic serialization. It is not good enough for default implementation but works.
``` C#
[AttributeUsage(AttributeTargets.Class, AllowMultiple = true)]
public class JsonKnownTypeAttribute : Attribute
{
    public Type NestedType { get; }
    public string Discriminator { get; }

    public JsonKnownTypeAttribute(Type nestedType, string discriminator)
    {
        NestedType = nestedType;
        Discriminator = discriminator;
    }
}

public class PolyJsonConverter : JsonConverterFactory
{
    private static readonly Type converterType = typeof(PolyJsonConverter<>);

    public override bool CanConvert(Type typeToConvert)
    {
        var attr = typeToConvert.GetCustomAttributes<JsonKnownTypeAttribute>(false);
        return attr.Any();
    }

    public override JsonConverter CreateConverter(Type typeToConvert, JsonSerializerOptions options)
    {
        var attr = typeToConvert.GetCustomAttributes<JsonKnownTypeAttribute>();
        var concreteConverterType = converterType.MakeGenericType(typeToConvert);
        return (JsonConverter)Activator.CreateInstance(concreteConverterType, attr);
    }
}

internal class PolyJsonConverter<T> : JsonConverter<T>
{
    private readonly Dictionary<string, Type> _discriminatorCache;
    private readonly Dictionary<Type, string> _typeCache;

    private static readonly JsonEncodedText TypeProperty = JsonEncodedText.Encode("$type");

    public PolyJsonConverter(IEnumerable<JsonKnownTypeAttribute> resolvers)
    {
        _discriminatorCache = resolvers.ToDictionary(p => p.Discriminator, p => p.NestedType);
        _typeCache = resolvers.ToDictionary(p => p.NestedType, p => p.Discriminator);
    }

    public override T Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        using var doc = JsonDocument.ParseValue(ref reader);
        if (!doc.RootElement.TryGetProperty(TypeProperty.EncodedUtf8Bytes, out var typeElement))
            throw new JsonException();

        var discriminator = typeElement.GetString();
        if (discriminator is null || !_discriminatorCache.TryGetValue(discriminator, out var type))
            throw new JsonException();


        return (T)doc.Deserialize(type, options);
    }

    public override void Write(Utf8JsonWriter writer, T value, JsonSerializerOptions options)
    {
        var type = value.GetType();
        if (!_typeCache.TryGetValue(type, out var discriminator))
            throw new JsonException();

        writer.WriteStartObject();
        writer.WritePropertyName(TypeProperty.EncodedUtf8Bytes);
        writer.WriteStringValue(discriminator);
        using var doc = JsonSerializer.SerializeToDocument(value, type, options);
        foreach (var prop in doc.RootElement.EnumerateObject())
            prop.WriteTo(writer);

        writer.WriteEndObject();
    }
}
```
Example of usage

``` C#
var data = new A[] { new B { PropA = "A-B", PropB = "B-B" }, new C { PropA = "A-C", PropC = "C-C" } };
var options = new JsonSerializerOptions
{
    Converters = { new PolyJsonConverter() }
};
var json = JsonSerializer.Serialize(data, options);
// 
var l = JsonSerializer.Deserialize<A[]>(json, options);

[JsonKnownType(typeof(B), "TypeB")]
[JsonKnownType(typeof(C), "TypeC")]
abstract class A
{
    public string PropA { get; set; }
}

class B : A
{
    public string PropB { get; set; }
}

class C : A
{
    public string PropC { get; set; }
}
```

This code produces following json:

``` javascript
[
  {
    "$type": "TypeB",
    "PropB": "B-B",
    "PropA": "A-B"
  },
  {
    "$type": "TypeC",
    "PropC": "C-C",
    "PropA": "A-C"
  }
]
```

**Comment by Symbai** [comment]
I would prefer an official solution rather than custom "workaround" as this might break in the future. If you are in control over both serialization and deserialization this could work. If you don't you risk that whatever you released into the world, will break your application sooner or later. I have the feeling many developers might agree with me. This "feature request" has been delayed over and over again. With 76 upvotes I really hope the .NET team finally gives it the attention it deserves.

**Comment by tthiery** [comment]
Indeed. There are enough samples on the internet. We need a canon solution. Polymorphic deserialization is a basic feature.

Another topic: Using attribute based metadata is hopefully not the only way. Not everyone is in control of the target types. I understand the security implications of providing callback lookups or list based inputs which can be misused by a stack overflow sample but attributes only a not a way (see EF Core and friends)

**Comment by eiriktsarpalis** [comment]
Given that the scope of this feature has changed substantially since the issue was originally opened, I've created a new [user story](https://github.com/dotnet/runtime/issues/63747) which should hopefully serve to better establish expectations wrt the brand of polymorphism we are planning to deliver. I'm going to close this one, feel free to continue the conversation in the new story if you have any feedback. Thanks!
